File: Sample/java/src/main/java/net/nend/sample/java/nativead/NativeMultipleLayoutListActivity.java
package net.nend.sample.java.nativead;

import android.app.ListActivity;
import android.content.Context;
import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import net.nend.android.NendAdNative;
import net.nend.android.NendAdNativeClient;
import net.nend.android.NendAdNativeViewBinder;
import net.nend.android.NendAdNativeViewHolder;
import net.nend.sample.java.R;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

public class NativeMultipleLayoutListActivity extends ListActivity {

    private final int NORMAL = 0;
    private final int AD1 = 1;
    private final int AD2 = 2;
    private final String TAG = getClass().getSimpleName();
    // 広告を表示したポジションの一覧
    private List<Integer> mPositionList = new ArrayList<>();
    // 表示したポジションと広告を紐付けて保持
    private HashMap<Integer, NendAdNative> mLoadedAd = new HashMap<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ArrayList<String> list = new ArrayList<>();
        for (int i = 1; i < 100; i++) {
            list.add("item" + i);
        }

        NativeListAdapter adapter = new NativeListAdapter(this, 0, list);
        setListAdapter(adapter);
    }

    class NativeListAdapter extends ArrayAdapter<String> {

        private NendAdNativeClient mClient;
        private NendAdNativeViewBinder mBinder;

        public NativeListAdapter(Context context, int resource, List<String> list) {
            super(context, resource, list);
            mBinder = new NendAdNativeViewBinder.Builder()
                    .adImageId(R.id.ad_image)
                    .titleId(R.id.ad_title)
                    .promotionNameId(R.id.ad_promotion_name)
                    .prId(R.id.ad_pr, NendAdNative.AdvertisingExplicitly.SPONSORED)
                    .build();

            mClient = new NendAdNativeClient(context, 485516, "16cb170982088d81712e63087061378c71e8aa5c");
        }

        @Override
        public int getViewTypeCount() {
            return 3;
        }

        @Override
        public int getItemViewType(int position) {
            if (position != 0 && position % 5 == 0) {
                return (position % 10 == 0) ? AD2 : AD1;
            }
            return NORMAL;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            final ViewHolder holder;
            final NendAdNativeViewHolder adHolder;
            switch (getItemViewType(position)) {
                case NORMAL:
                    if (convertView == null) {
                        convertView = LayoutInflater.from(getContext()).inflate(R.layout.native_list_row, parent, false);
                        holder = new ViewHolder();
                        holder.textView = convertView.findViewById(R.id.title);
                        holder.imageView = convertView.findViewById(R.id.thumbnail);
                        convertView.setTag(holder);
                    } else {
                        holder = (ViewHolder) convertView.getTag();
                    }
                    holder.textView.setText(getItem(position));
                    holder.imageView.setBackgroundColor(Color.LTGRAY);
                    break;
                case AD1:
                    if (mLoadedAd.containsKey(position)) {
                        adHolder = (NendAdNativeViewHolder) convertView.getTag();
                        mLoadedAd.get(position).intoView(adHolder);
                        break;
                    } else {
                        convertView = LayoutInflater.from(getContext()).inflate(R.layout.native_ad_left_row, parent, false);
                        loadAndSetAd(convertView, position);
                    }
                    break;
                case AD2:
                    if (mLoadedAd.containsKey(position)) {
                        adHolder = (NendAdNativeViewHolder) convertView.getTag();
                        mLoadedAd.get(position).intoView(adHolder);
                        break;
                    } else {
                        convertView = LayoutInflater.from(getContext()).inflate(R.layout.native_ad_right_row, parent, false);
                        loadAndSetAd(convertView, position);
                    }
                    break;
            }
            return convertView;
        }

        private void loadAndSetAd(View convertView, final int position) {

            final NendAdNativeViewHolder adHolder = mBinder.createViewHolder(convertView);
            convertView.setTag(adHolder);

            mClient.loadAd(new NendAdNativeClient.Callback() {
                @Override
                public void onSuccess(NendAdNative nendAdNative) {
                    Log.i(TAG, "広告取得成功");
                    mLoadedAd.put(position, nendAdNative);
                    mPositionList.add(position);
                    mLoadedAd.get(position).intoView(adHolder);
                }

                @Override
                public void onFailure(NendAdNativeClient.NendError nendError) {
                    Log.i(TAG, "広告取得失敗 " + nendError.getMessage());
                    // すでに取得済みの広告があればランダムで表示
                    if (!mLoadedAd.isEmpty()) {
                        Collections.shuffle(mPositionList);
                        mLoadedAd.get(mPositionList.get(0)).intoView(adHolder);
                    }
                }
            });
        }

        class ViewHolder {
            TextView textView;
            ImageView imageView;
        }
    }
}

File: Sample/kotlin/src/main/java/net/nend/sample/kotlin/nativead/NativeMultipleLayoutListActivity.kt
package net.nend.sample.kotlin.nativead

import android.app.ListActivity
import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.TextView
import net.nend.android.NendAdNative
import net.nend.android.NendAdNativeClient
import net.nend.android.NendAdNativeViewBinder
import net.nend.android.NendAdNativeViewHolder
import net.nend.sample.kotlin.R
import net.nend.sample.kotlin.nativead.NativeSampleActivity.Companion.NATIVE_API_KEY_SMALL_SQUARE
import net.nend.sample.kotlin.nativead.NativeSampleActivity.Companion.NATIVE_LOG_TAG
import net.nend.sample.kotlin.nativead.NativeSampleActivity.Companion.NATIVE_SPOT_ID_SMALL_SQUARE

class NativeMultipleLayoutListActivity : ListActivity() {

    private enum class ViewType(val id: Int) {
        NORMAL(0),
        AD1(1),
        AD2(2)
    }

    // 広告を表示したポジションの一覧
    private val positionList = mutableListOf<Int>()
    // 表示したポジションと広告を紐付けて保持
    private val loadedAd = mutableMapOf<Int, NendAdNative>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val list = (1..99).map { i -> "item$i" }
        listAdapter = NativeListAdapter(this, 0, list)
    }

    internal inner class NativeListAdapter(context: Context, resource: Int, list: List<String>) :
            ArrayAdapter<String>(context, resource, list) {

        private val client: NendAdNativeClient = NendAdNativeClient(context,
                NATIVE_SPOT_ID_SMALL_SQUARE, NATIVE_API_KEY_SMALL_SQUARE)
        private val binder: NendAdNativeViewBinder = NendAdNativeViewBinder.Builder()
                .adImageId(R.id.ad_image)
                .titleId(R.id.ad_title)
                .promotionNameId(R.id.ad_promotion_name)
                .prId(R.id.ad_pr, NendAdNative.AdvertisingExplicitly.SPONSORED)
                .build()

        override fun getViewTypeCount() = ViewType.values().size

        override fun getItemViewType(position: Int): Int {
            return if (position != 0 && position % 5 == 0) {
                if (position % 10 == 0) ViewType.AD2.id else ViewType.AD1.id
            } else ViewType.NORMAL.id
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val view: View
            val holder: ViewHolder
            val adHolder: NendAdNativeViewHolder

            when (getItemViewType(position)) {
                ViewType.AD1.id -> {
                    if (loadedAd.containsKey(position)) {
                        adHolder = convertView?.tag as NendAdNativeViewHolder
                        loadedAd[position]?.intoView(adHolder)
                        view = convertView
                    } else {
                        view = LayoutInflater
                                .from(context).inflate(R.layout.native_ad_left_row, parent, false)
                        loadAndSetAd(view, position)
                    }
                }
                ViewType.AD2.id -> {
                    if (loadedAd.containsKey(position)) {
                        adHolder = convertView?.tag as NendAdNativeViewHolder
                        loadedAd[position]?.intoView(adHolder)
                        view = convertView
                    } else {
                        view = LayoutInflater
                                .from(context).inflate(R.layout.native_ad_right_row, parent, false)
                        loadAndSetAd(view, position)
                    }
                }
                else -> {
                    if (convertView == null) {
                        view = LayoutInflater
                                .from(context).inflate(R.layout.native_list_row, parent, false)
                        holder = ViewHolder().apply {
                            textView = view.findViewById(R.id.title) as TextView
                            imageView = view.findViewById(R.id.thumbnail) as ImageView
                            view.tag = this
                        }
                    } else {
                        view = convertView
                        holder = view.tag as ViewHolder
                    }

                    holder.textView?.text = getItem(position)
                    holder.imageView?.setBackgroundColor(Color.LTGRAY)
                }
            }

            return view
        }

        private fun loadAndSetAd(convertView: View, position: Int) {
            val adHolder = binder.createViewHolder(convertView)
            convertView.tag = adHolder
            client.loadAd(object : NendAdNativeClient.Callback {
                override fun onSuccess(nendAdNative: NendAdNative) {
                    Log.i(NATIVE_LOG_TAG, "広告取得成功")
                    nendAdNative.run {
                        positionList.add(position)
                        intoView(adHolder)
                        loadedAd[position] = this
                    }
                }

                override fun onFailure(nendError: NendAdNativeClient.NendError) {
                    Log.i(NATIVE_LOG_TAG, "広告取得失敗: ${nendError.message}")
                    // すでに取得済みの広告があればランダムで表示
                    if (!loadedAd.isEmpty()) {
                        positionList.shuffle()
                        loadedAd[positionList[0]]?.intoView(adHolder)
                    }
                }
            })
        }

        internal inner class ViewHolder {
            var textView: TextView? = null
            var imageView: ImageView? = null
        }
    }
}